File: JavaScript/AoC2025/days/day5.js
import fs from "fs"
import path from "path"

export default class Day5 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day5.txt");
        const data = fs.readFileSync(filePath, "utf-8");
        this.lines = data.split('\n').map(line => line.trimEnd());
        const parsedData = this.parseData();
        this.available = parsedData[0];
        this.ranges = parsedData[1];
    }

    printResults() {
        console.log("Day 5:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${this.part2()}`);
    }

    part1() {
        let total = 0;
        for (const a of this.available) {
            for (const r of this.ranges) {
                if (a >= r[0] && a <= r[1]) {
                    total++;
                    break;
                }
            }
        }
        return total;
    }

    part2() {
        let total = 0;
        for (const range of this.ranges) {
            total += range[1] - range[0];
        }
        return total;
    }

    parseData() {
        let ranges = [];
        const available = [];
        let isRange = true;
        for (const line of this.lines) {
            if (line === "") {
                isRange = false;
                continue;
            }
            if (isRange) {
                const range = line.split('-');
                ranges.push([Number(range[0]), Number(range[1]) + 1]);
            } else {
                available.push(Number(line));
            }
        }
        ranges.sort((a, b) => a[0] === b[0] ? a[1] - b[1] : a[0] - b[0]);
        ranges = this.mergeRanges(ranges);
        return [available, ranges];
    }

    mergeRanges(ranges) {
        const combinedRanges = [];
        let start = ranges[0][0];
        let end = ranges[0][1];

        for (let i = 1; i < ranges.length; i++) {
            let s = ranges[i][0];
            let e = ranges[i][1];

            if (s >= end) {
                combinedRanges.push([start, end]);
                start = s;
                end = e;
            } else {
                end = e > end ? e : end;
            }
        }
        combinedRanges.push([start, end]);
        return combinedRanges;
    }
}

File: CSharp/AoC2025/Days/Day5.cs
namespace AoC2025.Days;

public static class Day5
{
    public static void Solution(out int partOne, out long partTwo)
    {
        ParseData(out List<(long, long)> ranges, out List<long> available);
        Console.WriteLine(ranges.Count);
        partOne = PartOne(ranges, available);
        partTwo = PartTwo(ranges);
    }

    private static int PartOne(List<(long start, long end)> ranges, List<long> available)
    {
        return available
            .Count(num => 
                ranges.Any(r => num >= r.start && num < r.end));
    }

    private static long PartTwo(List<(long start, long end)> ranges)
    {
        return ranges
            .Select(range => range.end - range.start)
            .Sum();
    }

    // Data methods

    private static string[] GetData() => File.ReadAllLines("../../Data/2025/day5.txt");

    private static void ParseData(out List<(long start, long end)> ranges, out List<long> available)
    {
        ranges = [];
        available = [];

        string[] data = GetData();
        bool isRange = true;

        foreach (string s in data)
        {
            if (s == "")
            {
                isRange = false;
                continue;
            }

            if (isRange)
            {
                string[] range = s.Split('-');
                long start = Convert.ToInt64(range[0]);
                long end = Convert.ToInt64(range[1]) + 1; // For simple range calculation.
                ranges.Add((start, end));
            }
            else
            {
                available.Add(Convert.ToInt64(s));
            }          
        }
        ranges.Sort((a, b) => a.start.CompareTo(b.start));
        ranges = CombineRanges(ranges);
    }

    private static List<(long, long)> CombineRanges(List<(long, long)> ranges)
    {
        List<(long, long)> combinedRanges = [];
        long start = 0, end = 0;

        foreach (var (s, e) in ranges)
        {
            if (s > end)
            {
                combinedRanges.Add((start, end));
                start = s;
                end = e;
            }
            else
            {
                end = Math.Max(end, e);
            }
        }

        combinedRanges.Add((start, end));

        return combinedRanges;
    }
}